website/blog/how-to-troubleshoot-w-bisect.md

```diff
@@ -0,0 +1,407 @@
+authors: son
+date: 2025-02-18:12:00:00
+image: /img/blog/git-bisect.jpg
+tags: [bazel, git, engineering]
+---
+
+Upgrading Bazel and the related dependencies can sometimes lead to unexpected issues.
+These issues can range from build failures to runtime errors, and generally, they can be hard to troubleshoot.
+
+So today, we will discuss how to narrow down the root cause of build failures after a dependency upgrade using `git bisect`.
+
+![](../static/img/blog/git-bisect.jpg)
+
+<!-- truncate -->
+
+## Build failed after upgrading Bazel
+
+If you are like me, you would enjoy having the latest and greatest tools in your project.
+And the tool I use the most is Bazel, so as the recent `8.1.0` release came out, I decided to upgrade our BuildBuddy repository to use it.
+
+```bash
+$ echo '8.1.0' > .bazelversion
+
+$ bazel test --config=remote-minimal //...
+...
+ERROR: /root/workspace/output-base/external/bazel_tools/tools/build_defs/repo/http.bzl:137:45: An error occurred during the fetch of repository 'rules_cc+':
+   Traceback (most recent call last):
+        File "/root/workspace/output-base/external/bazel_tools/tools/build_defs/repo/http.bzl", line 137, column 45, in _http_archive_impl
+                download_info = ctx.download_and_extract(
+Error in download_and_extract: com.google.devtools.build.lib.remote.common.CacheNotFoundException: Missing digest: abc605dd850f813bb37004b77db20106a19311a96b2da1c92b789da529d28fe1/178823
+...
+```
+
+Whoops! It seems like the build failed after upgrading Bazel to `8.1.0`.
+Flipping back to `8.0.1` made the problem go away, which means the issue was introduced in Bazel `8.1.0`.
+
+What should we do next?
+
+#### Auto-bisect with bazelisk
+
+Lucky for us, the Bazel team has provided us with a magic flag inside `bazelisk` that can help us with this.
+
+[--bisect](https://github.com/bazelbuild/bazelisk?tab=readme-ov-file#--bisect) flag automatically lists the commits between the two Bazel releases and helps us bisect them.
+It works like this:
+
+```bash
+alias bazel=bazelisk
+
+# Usage:
+#   bazelisk --bisect=<good commit hash>..<bad commit hash> test //foo:bar_test
+$ bazelisk --bisect=8.0.1..8.1.0 --output_base=$(mktemp -d) test --config=remote-minimal //...
+```
+
+**Explanation**: The last-known good release for us was `8.0.1`, and the bad release was `8.1.0`. Bazelisk will grab the list of commits between 2 versions from the Github API and start a bisect. For each bisect commit, it will grab a pre-built Bazel binary of that version from a Google Cloud Storage bucket and run the test command accordingly. The `.bazelversion` file will be ignored in these runs.
+
+Because the issue was related to external dependencies download, we want to have a clean workspace for each test run to avoid local cache affecting the test result. Instead of running `bazel clean --expunge` manually, we leverage `--output_base=$(mktemp -d)` here to create a temporary directory for each bisect run instead.
```

Comment:
You could set `BAZELISK_CLEAN=1` instead.

Reply:
Nice call. I switched the blog post to use BAZELISK_CLEAN